book.md

```diff
@@ -495,17 +492,160 @@ probably include, or be followed by, a section on the various sanitizers
+
+This can be achieved by chaining gadgets to set the registers `x0`, `x1`,
+`x2`, and then returning to `execve` in the C library.
+
+Let's assume we have the following gadgets:
+
+  1. A gadget that loads `x0` and `x1` from the stack:
+```
+  gadget_x0_x1:
+    ldp x0, x1, [sp]
+    ldp x20, x19, [sp, #64]
+    ldp x29, x30, [sp, #32]
+    ldr x21, [sp, #48]
+    add sp, sp, #0x50
+    ret
+```
+
+ 2. A gadget that sets `x2` to zero, but also clears `x0` as a side-effect:
+```
+  gadget_x2:
+    mov x2, xzr
+    mov x0, x2
+    ldp x20, x19, [sp, #32]
+    ldp x29, x30, [sp]
+    ldr x21, [sp, #16]
+    add sp, sp, #0x30
+    ret
+```
+
+Both gadgets also clobber several uninteresting registers, but since
+`gadget_x2` also clears `x0`, it becomes clear that we should use a
+ROP chain that:
+
+1. Returns to `gadget_x2`, which sets `x2` to zero.
+2. Returns to `gadget_x0_x1`, which sets `x0` and `x1` to the desired values.
+3. Returns to `execve`.
+
+@fig:rop-control-flow shows this control flow.
+
+![ROP example control flow](img/rop-control-flow){ width=30% #fig:rop-control-flow }
+
+![ROP example fake call stack](img/rop-call-stack){ width=80% #fig:rop-call-stack }
+
+We can achieve this by constructing the fake call stack shown in
+@fig:rop-call-stack, where "Original frame" marks the frame in which the
+address of `gadget_x2` has replaced a saved return address that will be loaded
+and returned to in the future. As an alternative, an attacker could place this
+fake call stack somewhere else, for example on the heap, and use a primitive
+that changes the stack pointer's value instead. This is known as stack
+pivoting.
+
+Note that this fake call stack contains NULL bytes, even without considering
+the various return addresses, and would need to be adjusted if an attacker
+wanted to use an overflow that relies on NULL-terminated string operations.
+
+A question that comes up when looking at the stack diagram is "how do we
+know the addresses of these gadgets"? We will talk a bit more about this in
+the next section.
+
+ROP gadgets like the ones used here may be easy to identify by visual
+inspection of a disassembled binary, but it's common for attackers to use
+"gadget scanner" tools in order to discover large numbers of gadgets
+automatically. Such tools can also be useful to a compiler engineer working on
+a code reuse attack mitigation, as they can point out code sequences that
+should be protected and have been missed.
+
+Jump-oriented programming [@Bletsch2011] is a variation on ROP, where gadgets
+can also end in indirect branch instructions instead of return instructions.
+The attacker chains a number of such gadgets through a dispatcher gadget, which
+loads pointers one after another from an array of pointers, and branches to
+each one in return. The gadgets used must be set up so that they branch or
+return back to the dispatcher after they're done. This is demonstrated in
+@fig:jop.
+
+![JOP example](img/jop){ width=50% #fig:jop }
```

Comment:
I find the figure hard to understand without an explanation of what happens. Maybe writing the words you'd say when explaining to someone what to look at in the figure, in the figure caption, would be a good way to have that explanation available for readers?